docs/design/auth-managers.rst

```diff
@@ -0,0 +1,403 @@
+
+
+Problems
+--------
+
+#. Inconsistency on privacy level methods (``.public``, ``.protected``, ``.private``).
+   They return projects with the same privacy level as the method name plus *all the projects*
+   for that user. This ends up in a mixture between privacy levels.
+
+#. ``.api`` method is not prepared to support APIv3 permissions:
+
+   * allow request for details on public projects
+   * deny request for listing on all projects not owned by authed user
+
+#. Easy to make mistakes if not immerse on this concept by using
+   ``.objects.all`` or ``.objects.filter`` directly (overpassing our methods).
+
+#. ``_add_user_repos``, which is the main method to know what objects the user has access to,
+   depends on ``django-guardian`` and ``view_project`` permission.
+   This one is added on ``Project.save`` and ``UserForm.save``.
+
+
+Goals
+-----
+
+#. unify all auth checks in one place
+#. make these checks compatible between community and corporate
+#. allow contributors to follow the pattern
+#. reduce mistakes when checking permissions
+#. support APIv3 detail (public projects) and list (owned projects) permissions
+
+
+Changes proposed (ideas)
+------------------------
+
+
+:.api:
+
+   Introducing a new boolean argument ``detail`` allows us to return
+   all public projects when ``True`` and only owned projects when
+   ``False`` (listing).
+
+   By defaulting in ``False`` we can keep the current behavior and
+   make this backward compatible. Although, I don't like this default
+   since a small mistake will give permissions to all the public
+   projects.
+
+   - projects the user has access to (``_add_users_repo``)
+
+   .. code-block:: python
+
+      def api(self, user=None, detail=True):
+          if detail:
+              # Allow accesing details for specific project
+              return self.public(user)
+
+          # Disallow listing on other's people projects
+          queryset = self.none()
+          if user:
+              return self._add_user_repos(queryset, user)
+          return queryset
+
+
+   This allow us to specify the type of request the user is doing on the API (list, detail)
+   and return the proper projects on each case.
+
+   .. code-block:: python
+
+      def get_queryset(self):
+          queryset = super().get_queryset()
+          return queryset.api(user=self.request.user, detail=self.detail)
+
+   With this modification, if the user access
+
+   * ``/api/v3/projects/`` will only see their projects
+   * ``/api/v3/projects/pip/`` will see the detail resource of this public project (even if it's not owner)
+
+
+:.api:
+
+   Use a ``APIAuthMixin`` to create an abstraction for the needed
+   queries on the API:
+
+   * without modifying the managers
+
+     .. code-block:: python
+
+        class APIAuthMixin:
+            def get_queryset(self):
+               if self.detail:
+                   return queryset.public
+               else:
+                   return queryset.private
```

Comment:
I think I like this approach. I think my current view is:

* We should only have `public`, `private`, and `for_user` queryset methods that we support
* We should build mixins for all the cases that we currently care about, and apply them to the API & Dashboard views consistently
* If we come to places where we are repeating logic more than 2x, we can consider bringing them back into the querysets, but I think keeping the queryset managers simple and having the mixins hold the integration with the code feels best

The main value here is that the *integration* of the queryset managers into the code only happens in one place, instead of in each view. This should ensure that we have the same logic in all our views.